### DISCUSSION TITLE
Log File

### DISCUSSION THREAD

**Discussion by tboy1337** [masterking32/MasterHamsterKombatBot#147] [General]
Logging code still doesn't create a log file.

**Comment by tboy1337** [comment]
Chat GPT‑4o recommends:
```
# Logging configuration
LOG_LEVEL = logging.DEBUG
LOGFORMAT = "%(log_color)s[Master HamsterKombat Bot]%(reset)s[%(log_color)s%(levelname)s%(reset)s] %(asctime)s %(log_color)s%(message)s%(reset)s"

# Set up root logger
logging.basicConfig(level=LOG_LEVEL, format=LOGFORMAT, datefmt="%Y-%m-%d %H:%M:%S")

# Formatter for console output
formatter = ColoredFormatter(LOGFORMAT, "%Y-%m-%d %H:%M:%S")

# Console handler
console_handler = logging.StreamHandler()
console_handler.setLevel(LOG_LEVEL)
console_handler.setFormatter(formatter)

# File handler for logging to a file
file_handler = logging.FileHandler("output.log")
file_handler.setLevel(LOG_LEVEL)
file_formatter = logging.Formatter("[Master HamsterKombat Bot][%(levelname)s] %(asctime)s %(message)s", "%Y-%m-%d %H:%M:%S")
file_handler.setFormatter(file_formatter)

# Get logger and add handlers
log = logging.getLogger("pythonConfig")
log.setLevel(LOG_LEVEL)
log.addHandler(console_handler)
log.addHandler(file_handler)

# End of logging configuration
```

This is making a log file but there are some issues because of colorlog where its adding ANSI escape codes to the log file.

**Comment by tboy1337** [comment]
Chat GPT‑4o:
```Key Points:

    Separate Formatters:
        Console Formatter: console_formatter uses ColoredFormatter to add color codes for console output.
        File Formatter: file_formatter uses logging.Formatter without color codes for file output.

    Ensure Correct Formatter Assignment:
        Console Handler: Assign console_formatter to console_handler.
        File Handler: Assign file_formatter to file_handler.

    No Overlap: Make sure that ColoredFormatter is not mistakenly applied to the FileHandler.

By clearly separating the formatters and ensuring that ColoredFormatter is only used with the console handler, the file logs should no longer include unwanted ANSI escape codes like [0m.
```

**Comment by Fy0urM** [comment]
```
import logging
from colorlog import ColoredFormatter

# Logging configuration
LOG_LEVEL = logging.DEBUG
LOGFORMAT_CONSOLE = "%(log_color)s[Master HamsterKombat Bot]%(reset)s[%(log_color)s%(levelname)s%(reset)s] %(asctime)s %(log_color)s%(message)s%(reset)s"
LOGFORMAT_FILE = "[Master HamsterKombat Bot][%(levelname)s] %(asctime)s %(message)s"

# Set up root logger
logging.basicConfig(level=LOG_LEVEL, format=LOGFORMAT_FILE, datefmt="%Y-%m-%d %H:%M:%S")

# Formatter for console output
formatter_console = ColoredFormatter(LOGFORMAT_CONSOLE, "%Y-%m-%d %H:%M:%S")

# Console handler
console_handler = logging.StreamHandler()
console_handler.setLevel(LOG_LEVEL)
console_handler.setFormatter(formatter_console)

# File handler for logging to a file
file_handler = logging.FileHandler("output.log")
file_handler.setLevel(LOG_LEVEL)
file_formatter = logging.Formatter(LOGFORMAT_FILE, "%Y-%m-%d %H:%M:%S")
file_handler.setFormatter(file_formatter)

# Get logger and add handlers 
log = logging.getLogger("pythonConfig")
log.setLevel(LOG_LEVEL)
log.addHandler(console_handler)
log.addHandler(file_handler)

# End of logging configuration

# Test logging
log.error(f"Log test")
```

**Comment by Fy0urM** [comment]
fixing log duplicate in console

```python
import logging
from colorlog import ColoredFormatter

# Logging configuration
LOG_LEVEL = logging.DEBUG
LOGFORMAT_CONSOLE = "%(log_color)s[Master HamsterKombat Bot]%(reset)s[%(log_color)s%(levelname)s%(reset)s] %(asctime)s %(log_color)s%(message)s%(reset)s"
LOGFORMAT_FILE = "[Master HamsterKombat Bot][%(levelname)s] %(asctime)s %(message)s"

# Create a logger
log = logging.getLogger("pythonConfig")
log.setLevel(LOG_LEVEL)

# Console handler
console_handler = logging.StreamHandler()
console_handler.setLevel(LOG_LEVEL)
console_handler.setFormatter(ColoredFormatter(LOGFORMAT_CONSOLE, "%Y-%m-%d %H:%M:%S"))

# File handler for logging to a file
file_handler = logging.FileHandler("output.log")
file_handler.setLevel(LOG_LEVEL)
file_handler.setFormatter(logging.Formatter(LOGFORMAT_FILE, "%Y-%m-%d %H:%M:%S"))

# Add handlers to the logger
log.addHandler(console_handler)
log.addHandler(file_handler)

# End of logging configuration

# Test logging
log.error(f"Log test")
```

**Reply by tboy1337** [reply]
> fixing log duplicate in console
> 
> ```python
> import logging
> from colorlog import ColoredFormatter
> 
> # Logging configuration
> LOG_LEVEL = logging.DEBUG
> LOGFORMAT_CONSOLE = "%(log_color)s[Master HamsterKombat Bot]%(reset)s[%(log_color)s%(levelname)s%(reset)s] %(asctime)s %(log_color)s%(message)s%(reset)s"
> LOGFORMAT_FILE = "[Master HamsterKombat Bot][%(levelname)s] %(asctime)s %(message)s"
> 
> # Create a logger
> log = logging.getLogger("pythonConfig")
> log.setLevel(LOG_LEVEL)
> 
> # Console handler
> console_handler = logging.StreamHandler()
> console_handler.setLevel(LOG_LEVEL)
> console_handler.setFormatter(ColoredFormatter(LOGFORMAT_CONSOLE, "%Y-%m-%d %H:%M:%S"))
> 
> # File handler for logging to a file
> file_handler = logging.FileHandler("output.log")
> file_handler.setLevel(LOG_LEVEL)
> file_handler.setFormatter(logging.Formatter(LOGFORMAT_FILE, "%Y-%m-%d %H:%M:%S"))
> 
> # Add handlers to the logger
> log.addHandler(console_handler)
> log.addHandler(file_handler)
> 
> # End of logging configuration
> 
> # Test logging
> log.error(f"Log test")
> ```

This code still produces ANSI escape codes in the .log file.

**Reply by Fy0urM** [reply]
> This code still produces ANSI escape codes in the .log file.

It may be necessary to split logger to console and logger to file into different instances, create a separate method for logirovagii which will accept text for output to log, and call functions for logirovagii to console and to file, before calling the logirovagii file function it is necessary to process the received string and remove all occurrences of escape codes from it, but since logger has its own methods for each type of logirovagii such as error, info, warning, etc., you need to figure out how to implement it. This is what I can think of on the fly, I am not good with python and its libraries. It can probably be done in a simpler way than I have suggested.

**Reply by Fy0urM** [reply]
> This code still produces ANSI escape codes in the .log file.

I think I managed to make a variant that will triple your expectations and will allow you not to make global changes to the code

```python
import logging
import re
from colorlog import ColoredFormatter


# creating filtering class
class RemoveEscapeCodesFilter(logging.Filter):
    def filter(self, record):
        escape_code_pattern = re.compile(r'\x1B\[[0-?]*[ -/]*[@-~]') # regex to catch escape codes
        record.msg = escape_code_pattern.sub('', record.msg) # cleanup message
        if record.args:
            record.args = tuple(escape_code_pattern.sub('', str(arg)) for arg in record.args)
        return True

LOG_LEVEL = logging.DEBUG # logging configuration

# format for console with colors
LOGFORMAT_CONSOLE = "%(log_color)s[Master HamsterKombat Bot]%(reset)s[%(log_color)s%(levelname)s%(reset)s] %(asctime)s %(log_color)s%(message)s%(reset)s"
# format for file without colors
LOGFORMAT_FILE = "[Master HamsterKombat Bot][%(levelname)s] %(asctime)s %(message)s"

# creating logger
log = logging.getLogger("pythonConfig")
log.setLevel(LOG_LEVEL)

# console handler
console_handler = logging.StreamHandler()
console_handler.setLevel(LOG_LEVEL)
console_handler.setFormatter(ColoredFormatter(LOGFORMAT_CONSOLE, "%Y-%m-%d %H:%M:%S"))

# file handler
file_handler = logging.FileHandler("output.log")
file_handler.setLevel(LOG_LEVEL)
file_handler.setFormatter(logging.Formatter(LOGFORMAT_FILE, "%Y-%m-%d %H:%M:%S"))

# add filter to remove ascii escape codes from log message
file_handler.addFilter(RemoveEscapeCodesFilter())

# adding handlers to logger
log.addHandler(console_handler)
log.addHandler(file_handler)

log.debug("\033[1;32mDebug test with colors\033[0m")
log.info("\033[1;36mInfo test with colors\033[0m")
log.warning("\033[1;33mWarning test with colors\033[0m")
log.error("\033[1;31mError test with colors\033[0m")
log.critical("\033[1;35mCritical test with colors\033[0m")

```

**Reply by tboy1337** [reply]
> > This code still produces ANSI escape codes in the .log file.
> 
> I think I managed to make a variant that will triple your expectations and will allow you not to make global changes to the code
> 
> ```python
> import logging
> import re
> from colorlog import ColoredFormatter
> 
> 
> # creating filtering class
> class RemoveEscapeCodesFilter(logging.Filter):
>     def filter(self, record):
>         escape_code_pattern = re.compile(r'\x1B\[[0-?]*[ -/]*[@-~]') # regex to catch escape codes
>         record.msg = escape_code_pattern.sub('', record.msg) # cleanup message
>         if record.args:
>             record.args = tuple(escape_code_pattern.sub('', str(arg)) for arg in record.args)
>         return True
> 
> LOG_LEVEL = logging.DEBUG # logging configuration
> 
> # format for console with colors
> LOGFORMAT_CONSOLE = "%(log_color)s[Master HamsterKombat Bot]%(reset)s[%(log_color)s%(levelname)s%(reset)s] %(asctime)s %(log_color)s%(message)s%(reset)s"
> # format for file without colors
> LOGFORMAT_FILE = "[Master HamsterKombat Bot][%(levelname)s] %(asctime)s %(message)s"
> 
> # creating logger
> log = logging.getLogger("pythonConfig")
> log.setLevel(LOG_LEVEL)
> 
> # console handler
> console_handler = logging.StreamHandler()
> console_handler.setLevel(LOG_LEVEL)
> console_handler.setFormatter(ColoredFormatter(LOGFORMAT_CONSOLE, "%Y-%m-%d %H:%M:%S"))
> 
> # file handler
> file_handler = logging.FileHandler("output.log")
> file_handler.setLevel(LOG_LEVEL)
> file_handler.setFormatter(logging.Formatter(LOGFORMAT_FILE, "%Y-%m-%d %H:%M:%S"))
> 
> # add filter to remove ascii escape codes from log message
> file_handler.addFilter(RemoveEscapeCodesFilter())
> 
> # adding handlers to logger
> log.addHandler(console_handler)
> log.addHandler(file_handler)
> 
> log.debug("\033[1;32mDebug test with colors\033[0m")
> log.info("\033[1;36mInfo test with colors\033[0m")
> log.warning("\033[1;33mWarning test with colors\033[0m")
> log.error("\033[1;31mError test with colors\033[0m")
> log.critical("\033[1;35mCritical test with colors\033[0m")
> ```

@Fy0urM its working at about 90 percent but a few times ive seen it error, hes one of them, any ideas?
```
[Master HamsterKombat Bot][INFO] 2024-08-27 02:03:55 [Account 10] Best upgrade is Сoindesk with profit 193 and price 159.84k, Level: 14
--- Logging error ---
Traceback (most recent call last):
  File "C:\Program Files\Python312\Lib\logging\__init__.py", line 1163, in emit
    stream.write(msg + self.terminator)
  File "C:\Program Files\Python312\Lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeEncodeError: 'charmap' codec can't encode character '\u0421' in position 85: character maps to <undefined>
Call stack:
  File "C:\Users\FARM\Downloads\Hamster Kombat\MasterHamsterKombatBot-main-7\main.py", line 1554, in <module>
    main()
  File "C:\Users\FARM\Downloads\Hamster Kombat\MasterHamsterKombatBot-main-7\main.py", line 1548, in main
    asyncio.run(RunAccounts())
  File "C:\Users\FARM\Downloads\Hamster Kombat\MasterHamsterKombatBot-main-7\main.py", line 1515, in RunAccounts
    account.Start()
  File "C:\Users\FARM\Downloads\Hamster Kombat\MasterHamsterKombatBot-main-7\main.py", line 1409, in Start
    if not self.BuyBestCard():
  File "C:\Users\FARM\Downloads\Hamster Kombat\MasterHamsterKombatBot-main-7\main.py", line 718, in BuyBestCard
    log.info(
Message: '[Account 10] Best upgrade is Сoindesk with profit 193 and price 159.84k, Level: 14'
Arguments: ()
[Master HamsterKombat Bot][INFO] 2024-08-27 02:03:55 [Account 10] Attempting to buy the best card...
```

**Reply by Fy0urM** [reply]
> ```python
> file_handler = logging.FileHandler("output.log")
> ```

try changing this line like this
```python
file_handler = logging.FileHandler("output.log", encoding='utf-8')
```

**Reply by Fy0urM** [reply]
> [Master HamsterKombat Bot][INFO] 2024-08-27 02:03:55 [Account 10] Best upgrade is Сoindesk with profit 193 and price 159.84k, Level: 14

it looks very funny, if I understood correctly in `Coindesk` the first character `C` is written in Cyrillic instead of Latin for some reason

**Reply by tboy1337** [reply]
> > [Master HamsterKombat Bot][INFO] 2024-08-27 02:03:55 [Account 10] Best upgrade is Сoindesk with profit 193 and price 159.84k, Level: 14
> 
> it looks very funny, if I understood correctly in `Coindesk` the first character `C` is written in Cyrillic instead of Latin for some reason

Thanks for the code update. Yea that is strange about the cyrillic character, i'm not sure how you can tell that by looking at it lol.

**Reply by Fy0urM** [reply]
> character '\u0421' in

`\u0421` that С in cyrillic

**Reply by tboy1337** [reply]
> > character '\u0421' in
> 
> `\u0421` that С in cyrillic

Ah ok, thanks.

**Comment by tboy1337** [comment]
https://github.com/masterking32/MasterHamsterKombatBot/pull/241

**Comment by tboy1337** [comment]
If anyone wants local logging use [main.py](https://github.com/tboy1337/MasterHamsterKombatBot/blob/test/main.py) in my test branch.

**Comment by tboy1337** [comment]
Moved to https://github.com/masterking32/MasterHamsterKombatBot/discussions/301.
